### PR TITLE
Delete conversation

### PR DESCRIPTION
**Pull request by lee-tammy**
Let me know if I should fix anything before I merge into the master.

### CODE REVIEW

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -370,6 +373,29 @@ public void invoke(List<String> args) {
           }
         });
 
+    panel.register("c-remove", new Panel.Command(){
+      @Override
+      public void invoke(List<String> args){
+        final String name = args.size() > 0 ? args.get(0).trim() : "";
```
Comment (alanhzhao) [inline src/codeu/chat/client/commandline/Chat.java]: Prefer isEmpty() over comparing size() to 0

File: src/codeu/chat/client/commandline/Chat.java [hunk 2]
```diff
@@ -370,6 +373,29 @@ public void invoke(List<String> args) {
...
+        if(name.length() > 0){
```
Comment (alanhzhao) [inline src/codeu/chat/client/commandline/Chat.java]: Prefer the following, which is more readable:

    if (name.isEmpty()) {
      System.out.println("ERROR: Wrong format");
      return;
    }

Generally, it's good to minimize the number of indentation levels, and put related code next as close as possible. So if we're effectively returning early with an error, express it this way. Also, this error can be more descriptive--maybe "no conversation specified"?

File: src/codeu/chat/client/commandline/Chat.java [hunk 3]
```diff
@@ -398,7 +424,8 @@ public void invoke(List<String> args) {
             }
           }
         });
-
+
```
Comment (alanhzhao) [inline src/codeu/chat/client/commandline/Chat.java]: Nit: unnecessary whitespace and line break

File: src/codeu/chat/client/core/Controller.java [hunk 4]
```diff
@@ -95,7 +95,6 @@ public User newUser(String name) {
     return response;
   }
 
-  @Override
```
Comment (alanhzhao) [inline src/codeu/chat/client/core/Controller.java]: Probably want to keep this `@Override`.

File: src/codeu/chat/client/core/Controller.java [hunk 5]
```diff
@@ -120,6 +119,19 @@ public ConversationHeader newConversation(String title, Uuid owner, UserType def
     return response;
   }
 
+  public void removeConversation(ConversationHeader conversation){
```
Comment (alanhzhao) [inline src/codeu/chat/client/core/Controller.java]: Also add `@Override` here.

File: src/codeu/chat/common/NetworkCode.java [hunk 6]
```diff
@@ -31,6 +31,7 @@
       NEW_USER_RESPONSE = 12,
       NEW_CONVERSATION_REQUEST = 13,
       NEW_CONVERSATION_RESPONSE = 14,
+      REMOVE_CONVERSATION_REQUEST = 15,
```
Comment (alanhzhao) [inline src/codeu/chat/common/NetworkCode.java:34]: Do we need a corresponding response?

File: src/codeu/chat/server/Controller.java [hunk 7]
```diff
@@ -195,6 +195,11 @@ public ConversationHeader newConversation(Uuid id, String title, Uuid owner, Tim
     return conversation;
   }
 
+  public void removeConversation(ConversationHeader conversation){
+    model.remove(conversation);
+  }
+    
+
```
Comment (alanhzhao) [inline src/codeu/chat/server/Controller.java]: Nit: extra line break

File: src/codeu/chat/client/core/Controller.java [hunk 8]
```diff
@@ -120,6 +119,19 @@ public ConversationHeader newConversation(String title, Uuid owner, UserType def
...
+    try(final Connection connection = source.connect()){
+
+      Serializers.INTEGER.write(connection.out(), NetworkCode.REMOVE_CONVERSATION_REQUEST);
+      Serializers.NULLABLE(ConversationHeader.SERIALIZER).write(connection.out(), conversation);
+    }catch(Exception ex){
+      LOG.error(ex, "Exception during call on server.");
+    }
+        
+  }
+
```
Comment (alanhzhao) [inline src/codeu/chat/client/core/Controller.java]: Remove line breaks causing empty lines in 125, 131, and 133. Empty lines should make sense semantically.

File: src/codeu/chat/server/Server.java [hunk 9]
```diff
@@ -149,6 +149,17 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
           }
         });
 
+    this.commands.put(NetworkCode.REMOVE_CONVERSATION_REQUEST, new Command(){
+      @Override
+      public void onMessage(InputStream in, OutputStream out)throws IOException{
+       
+        final ConversationHeader conversation = 
+            Serializers.NULLABLE(ConversationHeader.SERIALIZER).read(in);
+
```
Comment (alanhzhao) [inline src/codeu/chat/server/Server.java]: remove empty lines 155, 158. This method is short, so no need for them.

**Review by alanhzhao — CHANGES_REQUESTED** [review]

File: src/codeu/chat/common/NetworkCode.java (hunk 6 above)
Comment (lee-tammy) [reply src/codeu/chat/common/NetworkCode.java:34]: Is it required there be a response if the user shouldn't be expected anything after making the request? Maybe a message saying if the conversation was created successfully but I don't think that would need a response.

Comment (alanhzhao) [reply src/codeu/chat/common/NetworkCode.java:34]: Since there's REMOVE_USER_RESPONSE and REMOVE_INTEREST_RESPONSE in [Server.java](https://github.com/lee-tammy/CodeU-Summer-2017/pull/11/files/7378526d6af24c4a287d81a413c8bcc0e31c71ab#diff-e2387988b5a78207609d160642f28cd3), I think it would make sense to add it.

Comment (lee-tammy) [reply src/codeu/chat/common/NetworkCode.java:34]: Okay, that makes sense. I added it in. Let me know if I need to anything else in.

**Review by Weech4theStars — APPROVED** [review]
looks good to me!

File: src/codeu/chat/server/Server.java [hunk 10]
```diff
@@ -149,6 +149,16 @@ public void onMessage(InputStream in, OutputStream out) throws IOException {
           }
         });
 
+    this.commands.put(NetworkCode.REMOVE_CONVERSATION_REQUEST, new Command(){
+      @Override
+      public void onMessage(InputStream in, OutputStream out)throws IOException{ 
+        final ConversationHeader conversation = 
+            Serializers.NULLABLE(ConversationHeader.SERIALIZER).read(in);
+        controller.removeConversation(conversation);
+        Serializers.INTEGER.write(out, NetworkCode.REMOVE_CONVERSATION_REQUEST);
```
Comment (alanhzhao) [inline src/codeu/chat/server/Server.java]: REMOVE_CONVERSATION_RESPONSE

**Review by alanhzhao — APPROVED** [review]
Approved, conditional on 1 change

**Comment by adsnaider** [comment]
Looks good to me. Just fix the REMOVE_CONVERSATION_RESPONSE that Alan pointed out and that should be it.
